File: code/ex949.java
class Solution {
    public String largestTimeFromDigits(int[] A) {
        int n = A.length;
        String res = "";
        int max = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    for (int l = 0; l < n; l++) {
                        if (i == j || i == k || i == l || j == k || j == l || k == l)
                            continue;
                        if (valid(A[i], A[j], A[k], A[l])) {
                            if (A[l] + A[k] * 10 + A[j] * 100 + A[i] * 1000 >= max) {
                                max = A[l] + A[k] * 10 + A[j] * 100 + A[i] * 1000;
                                res = String.valueOf(A[i]) + String.valueOf(A[j]) + ":" + String.valueOf(A[k]) + String.valueOf(A[l]);
                            }
                        }
                    }
                }
            }
        }
        
        return res;
    }
    
    private boolean valid(int i, int j, int k, int l) {
        if (i > 2 || k > 5)    return false;
        if (i == 2 && j > 3)    return false;
        return true;
    }
}

File: code/ex949.cpp
class Solution {
    string MAX = "";
public:
    string largestTimeFromDigits(vector<int>& A) {
        vector<bool> visited(4);
        dfs(A, visited, 0, "");
        if (MAX == "")    return "";
        MAX.insert(MAX.begin() + 2, ':');
        return MAX;
    }
    
private:
    void dfs(vector<int>& A, vector<bool>& visited, int s, string cur) {
        if (s == 4) {
            if (valid(cur))    MAX = max(MAX, cur);
            return;
        }
        
        for (int i = 0; i < 4; i++) {
            if (visited[i])    continue;
            visited[i] = 1;
            dfs(A, visited, s + 1, cur + to_string(A[i]));
            visited[i] = 0;
        }
    }
    
    bool valid(string s) {
        int d1 = s[0] - '0', d2 = s[1] - '0', d3 = s[2] - '0', d4 = s[3] - '0';
        if (d1 > 2 || (d1 == 2 && d2 >= 4))    return 0;
        return d3 <= 5;
    }
};